cpp/memilio/io/README.md

```diff
@@ -140,14 +140,48 @@ To set the value of a parameter from the command line, first type the correspond
+The History Object is providing a way to save data throughout the simulation process. It offers an interface where users can define the data to be saved from a given object using loggers and the method of saving it using writers. Afterward, the user can access this data from the History Object and manipulate it. For a basic logger use case, refer to [this example](../../examples/history.cpp). For an example demonstrating using a logger in the ABM, refer to [this example](../../examples/abm_history_example.cpp).
 
-- The `Writer` defines the type `Data` to store all records (i.e., the return values of `Logger::log`), and the function `template <class Logger> static void add_record(const Logger::Type&, Data&)` to add a new record. `add_record` is used whenever `History::log` was called and `Logger::should_log` is true.
+### Loggers
 
-- The `History` class is templated on the `Writer` that is used to handle the data (e.g., store it into an array), and the `Loggers` that are used to log data. The `Loggers` used for a `History` must be unique.
+The `Logger` struct is a tool for logging data from a given object. Each user-implemented logger must have a `Type` and implement two functions: `Type log(const T&)` and `bool should_log(const T&)`. The input `T` for these functions is the same as the one given to the `History` member-function `History::log`.
+
+- `Type`: Return Type of `log`.
+- `log`: This function determines which data from the input `T` is saved. It must have the same return Type `Type` as the loggers Type `Type`.
+- `should_log`: This function must return a boolean to determine if data should be logged and can use the input `T` for this, e.g. if `T` fullfills some criteria.
+
+Users can derive their loggers from `LogOnce` or `LogAlways` to use a predefined `should_log` function. `LogOnce` logs only at the first call of `log`, while `LogAlways` logs every time `log` is called. All implemented loggers must have unique types and be constructible/destructible. For user-defined examples in the ABM, refer to [this file](../../models/abm/common_abm_loggers.h).
+
+### Writers
+
+The `Writer` struct defines how to store the logged data from one or more implemented `Loggers`. Each user-implemented `Writer` must have a `Data` Type and implement the `static void add_record(const typename Logger::Type& t, Data& data)` function.
+
+- `Data`: This is a `std::tuple` of classes which store the data returned by the Loggers. For example, this can be Vectors or TimeSeries.
+- `add_record`. This should manipulate the data of the Writer to store the data returned by the loggers. It is used whenever `History::log` was called and `Logger::should_log` is true.
```

Comment:
It is not the "data of the Writer", the data is passed as an argument.

"whenever History::log **is** called"